### PR TITLE
change PM logic for summary

### PR DESCRIPTION
**Pull request by Seeker1911**

### CODE REVIEW

File: rest-api/dao/physical_measurements_dao.py [hunk 1]
```diff
@@ -291,22 +291,28 @@ def _update_participant_summary(self, session, obj):
...
+      # new PM or if a PM was restored, it is complete again.
       participant_summary.physicalMeasurementsStatus = PhysicalMeasurementsStatus.COMPLETED
+      participant_summary.physicalMeasurementsTime = obj.created
+
+    # these fields set on all new measurements and cancelled measurements that don't have a
+    # previous good measurement.
+    if obj.status != PhysicalMeasurementsStatus.CANCELLED or \
+        (obj.status == PhysicalMeasurementsStatus.CANCELLED and not
+         self.has_uncancelled_pm(session, participant)):
+      participant_summary.physicalMeasurementsFinalizedTime = obj.finalized
```
Comment (calbach) [inline rest-api/dao/physical_measurements_dao.py]: Per offline discussion, my understanding of what Healthpro wants here is the following:


```
latest = self.find_latest_uncancelled_pm(...)
ps.fintalizedTime = latest.finalized
ps.siteId = latest.siteId
...
``

@shyamnaru  can you verify?

File: rest-api/dao/physical_measurements_dao.py [hunk 2]
```diff
@@ -291,28 +291,45 @@ def _update_participant_summary(self, session, obj):
...
+      participant_summary.physicalMeasurementsFinalizedSiteId = get_latest_pm.finalizedSiteId
+
 
     participant_summary_dao.update_enrollment_status(participant_summary)
     session.merge(participant_summary)
 
     return participant_summary
 
+  def get_latest_pm(self, session, participant):
+    query = session.query(PhysicalMeasurements).filter(PhysicalMeasurements.status.isnot(
```
Comment (calbach) [inline rest-api/dao/physical_measurements_dao.py]: nit: I'm not a SQLAlchemy expert, but maybe put the participant filter first? It probably won't matter but I'm slightly scared this would do something like pull all uncancelled PM into memory first.

File: rest-api/dao/physical_measurements_dao.py [hunk 3]
```diff
@@ -291,28 +291,45 @@ def _update_participant_summary(self, session, obj):
...
+      PhysicalMeasurementsStatus.CANCELLED)).filter_by(
+      participantId=participant.participantId).order_by(
+      PhysicalMeasurements.finalized.desc()).first()
+    return query
```
Comment (calbach) [inline rest-api/dao/physical_measurements_dao.py]: nit: no need for variable, just return above

Comment (Seeker1911) [reply rest-api/dao/physical_measurements_dao.py]: done

File: rest-api/dao/physical_measurements_dao.py (hunk 2 above)
Comment (Seeker1911) [reply rest-api/dao/physical_measurements_dao.py]: done

File: rest-api/dao/physical_measurements_dao.py [hunk 4]
```diff
@@ -291,32 +291,49 @@ def _update_participant_summary(self, session, obj):
...
+    # These fields set on any measurement not cancelled
+    elif obj.status != PhysicalMeasurementsStatus.CANCELLED:
+      # new PM or if a PM was restored, it is complete again.
       participant_summary.physicalMeasurementsStatus = PhysicalMeasurementsStatus.COMPLETED
+      participant_summary.physicalMeasurementsTime = obj.created
+      participant_summary.physicalMeasurementsFinalizedTime = obj.finalized
+      participant_summary.physicalMeasurementsCreatedSiteId = obj.createdSiteId
+      participant_summary.physicalMeasurementsFinalizedSiteId = obj.finalizedSiteId
+
+    if obj.status and obj.status == PhysicalMeasurementsStatus.CANCELLED and \
```
[the remark this answers is not in the document]
Comment (Seeker1911) [reply rest-api/dao/physical_measurements_dao.py]: made elif, we do need it because we changed these fields to None when cancelled, but if they have a previous good PM we don't want participant summary to reflect these changes.

File: rest-api/dao/physical_measurements_dao.py [hunk 5]
```diff
@@ -291,32 +291,49 @@ def _update_participant_summary(self, session, obj):
...
+      participant_summary.physicalMeasurementsFinalizedSiteId = get_latest_pm.finalizedSiteId
 
     participant_summary_dao.update_enrollment_status(participant_summary)
     session.merge(participant_summary)
 
     return participant_summary
 
+  def get_latest_pm(self, session, participant):
+    return session.query(PhysicalMeasurements).filter_by(
+      participantId=participant.participantId).filter(PhysicalMeasurements.finalized !=
```
[the remark this answers is not in the document]
Comment (Seeker1911) [reply rest-api/dao/physical_measurements_dao.py]: don't think so, I'm looking for measurements which have been finalized specifically.

File: rest-api/dao/physical_measurements_dao.py [hunk 6]
```diff
@@ -291,32 +291,49 @@ def _update_participant_summary(self, session, obj):
...
+                                                      None).order_by(
+                                                      PhysicalMeasurements.finalized.desc()).first()
+
   def has_uncancelled_pm(self, session, participant):
     """return True if participant has at least one physical measurement that is not cancelled"""
     query = session.query(PhysicalMeasurements.status).filter_by(
-      participantId=participant.participantId).all()
+      participantId=participant.participantId).filter(PhysicalMeasurements.finalized !=
```
[the remark this answers is not in the document]
Comment (Seeker1911) [reply rest-api/dao/physical_measurements_dao.py]: same, unless I'm not understanding your question 🤷‍♂️

File: rest-api/dao/physical_measurements_dao.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (danqrodney) [reply rest-api/dao/physical_measurements_dao.py]: I guess all canceled PMs get their finalized field set to None here:

https://github.com/all-of-us/raw-data-repository/blob/46507b75215c5926fa3d556394f6351b650f29ab/rest-api/dao/physical_measurements_dao.py#L394

As long as this runs first, should be fine...  you may just want to add a comment about it.

File: rest-api/dao/physical_measurements_dao.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (danqrodney) [reply rest-api/dao/physical_measurements_dao.py]: OK, got it.

File: rest-api/dao/physical_measurements_dao.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Seeker1911) [reply rest-api/dao/physical_measurements_dao.py]: right so this finds not cancelled orders basically and in order of date finalized.

**Comment by Seeker1911** [comment]
A bug in prod created the following stacktrace:
` File "/base/data/home/apps/s~all-of-us-rdr-prod/v1-37-rc3.414038559755011735/dao/participant_summary_dao.py", line 449, in calculate_core_ordered_sample_time participant_summary, field_name_prefix='sampleOrderStatus') File "/base/data/home/apps/s~all-of-us-rdr-prod/v1-37-rc3.414038559755011735/dao/participant_summary_dao.py", line 473, in calculate_max_core_sample_time participant_summary.physicalMeasurementsFinalizedTime]) TypeError: can't compare datetime.datetime to  #NoneType`
 
The test I added was a proof of concept that hits the calculate sample logic in participant summary. It's a bit of an edge case. You have to have a cancelled PM, a new good PM, consent, PPI modules complete and then insert a new biobank order. When I made finalized time to None in physical measurements dao I get the same error as the stack trace above. As it is in this PR, the test passes.

File: rest-api/test/unit_test/api_test/biobank_order_api_test.py [hunk 7]
```diff
@@ -502,11 +512,69 @@ def test_not_pairing_at_pm_when_has_bio(self):
...
+    self._submit_consent_questionnaire_response(self.participant_id, questionnaire_id_1,
+                                                CONSENT_PERMISSION_YES_CODE, time=TIME_6)
+
+    self.submit_questionnaire_response(self.participant_id, questionnaire_id,
+                                       RACE_NONE_OF_THESE_CODE, None, None,
+                                       datetime.date(1978, 10, 10))
+
+    self._submit_empty_questionnaire_response(self.participant_id, questionnaire_id_2)
+
+    # send a biobank ordewr
```
Comment (calbach) [inline rest-api/test/unit_test/api_test/biobank_order_api_test.py]: nit: spelling

File: rest-api/dao/physical_measurements_dao.py [hunk 8]
```diff
@@ -291,32 +291,49 @@ def _update_participant_summary(self, session, obj):
...
-        participant_summary.physicalMeasurementsTime = None
+    # These fields set on measurement that is cancelled and doesn't have a previous good measurement
+    if obj.status and obj.status == PhysicalMeasurementsStatus.CANCELLED and not \
+       self.has_uncancelled_pm(session, participant):
 
-    if participant_summary.physicalMeasurementsStatus != PhysicalMeasurementsStatus.CANCELLED or \
-       obj.status != PhysicalMeasurementsStatus.CANCELLED:
-      # if a PM was restored, it is complete again.
+      participant_summary.physicalMeasurementsStatus = PhysicalMeasurementsStatus.CANCELLED
+      participant_summary.physicalMeasurementsTime = None
```
Comment (calbach) [inline rest-api/dao/physical_measurements_dao.py]: Should you be setting finalized time and siteId to `None` here?

File: rest-api/dao/physical_measurements_dao.py [hunk 9]
```diff
@@ -291,32 +291,49 @@ def _update_participant_summary(self, session, obj):
...
+      participant_summary.physicalMeasurementsFinalizedSiteId = get_latest_pm.finalizedSiteId
 
     participant_summary_dao.update_enrollment_status(participant_summary)
     session.merge(participant_summary)
 
     return participant_summary
 
+  def get_latest_pm(self, session, participant):
+    return session.query(PhysicalMeasurements).filter_by(
+      participantId=participant.participantId).filter(PhysicalMeasurements.finalized !=
```
Comment (calbach) [inline rest-api/dao/physical_measurements_dao.py]: nit: this indentation is pretty rough, can it be cleaned up?

File: rest-api/dao/physical_measurements_dao.py [hunk 10]
```diff
@@ -291,32 +291,49 @@ def _update_participant_summary(self, session, obj):
...
+                                                      None).order_by(
+                                                      PhysicalMeasurements.finalized.desc()).first()
+
   def has_uncancelled_pm(self, session, participant):
     """return True if participant has at least one physical measurement that is not cancelled"""
     query = session.query(PhysicalMeasurements.status).filter_by(
-      participantId=participant.participantId).all()
+      participantId=participant.participantId).filter(PhysicalMeasurements.finalized !=
```
Comment (calbach) [inline rest-api/dao/physical_measurements_dao.py]: I don't see why this change was necessary - are there going to be uncancelled PMs which don't have a finalized time?

File: rest-api/dao/physical_measurements_dao.py (hunk 8 above)
Comment (Seeker1911) [reply rest-api/dao/physical_measurements_dao.py]: Those fields (and most) are set in _do_update_with_patch on the measurement  object. In update_participant_summary I'm only changing fields specific to the participant summary.

File: rest-api/dao/physical_measurements_dao.py (hunk 9 above)
Comment (Seeker1911) [reply rest-api/dao/physical_measurements_dao.py]: done.

File: rest-api/dao/physical_measurements_dao.py (hunk 10 above)
Comment (Seeker1911) [reply rest-api/dao/physical_measurements_dao.py]: There can be. Not sent from HPRO but if another order was cancelled and finalized was set to None it could be returned here.

File: rest-api/test/unit_test/api_test/biobank_order_api_test.py (hunk 7 above)
Comment (Seeker1911) [reply rest-api/test/unit_test/api_test/biobank_order_api_test.py]: done

File: rest-api/dao/physical_measurements_dao.py (hunk 8 above)
Comment (calbach) [reply rest-api/dao/physical_measurements_dao.py]: I was referring to `physicalMeasurementsFinalizedTime` and `physicalMeasurementsFinalizedSiteId`

**Comment by Seeker1911** [comment]
physicalMeasurementsFinalizedTime and siteId are set in _do_update_with_patch. Merging.
